service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java

```diff
@@ -301,45 +326,91 @@ public void addComplyOnBehalfResponsesToDirectionsInStandardDirectionsOrder(Case
+        final UUID[] id = new UUID[1];
 
         return directions.stream()
             .map(directionElement -> {
-                id.set(directionElement.getId());
+                id[0] = directionElement.getId();
 
                 return directionElement.getValue().getResponses();
             })
             .flatMap(List::stream)
-            .map(element -> addCourtAssigneeAndDirectionId(id, element))
+            .map(element -> getDirectionResponse(eventId, authorisation, id[0], element, assignee))
             .collect(toList());
     }
 
-    private Element<DirectionResponse> addCourtAssigneeAndDirectionId(AtomicReference<UUID> id,
-                                                                      Element<DirectionResponse> element) {
+    private Element<DirectionResponse> getDirectionResponse(ComplyOnBehalfEvent event,
+                                                            String authorisation,
+                                                            UUID id,
+                                                            Element<DirectionResponse> response,
+                                                            DirectionAssignee assignee) {
+        if (event == COMPLY_ON_BEHALF_SDO) {
+            return addCourtAssigneeAndDirectionId(id, response);
+        } else {
+            return addResponderAssigneeAndDirectionId(response, authorisation, assignee, id);
+        }
+    }
+
+    //TODO: if name of user complying for court is added then this can be merged with logic from method below.
+    private Element<DirectionResponse> addCourtAssigneeAndDirectionId(UUID id, Element<DirectionResponse> element) {
         return Element.<DirectionResponse>builder()
             .id(element.getId())
             .value(element.getValue().toBuilder()
                 .assignee(COURT)
-                .directionId(id.get())
+                .directionId(id)
                 .build())
             .build();
     }
 
+    private Element<DirectionResponse> addResponderAssigneeAndDirectionId(Element<DirectionResponse> response,
+                                                                          String authorisation,
+                                                                          DirectionAssignee assignee,
+                                                                          UUID id) {
+        return Element.<DirectionResponse>builder()
+            .id(response.getId())
+            .value(response.getValue().toBuilder()
+                .assignee(assignee)
+                .responder(getUsername(response, authorisation))
+                .directionId(id)
+                .build())
+            .build();
+    }
+
+    private String getUsername(Element<DirectionResponse> element, String authorisation) {
+        if (isEmpty(element.getValue().getResponder())) {
+            return userDetailsService.getUserName(authorisation);
```

Comment:
I almost wished we had this authorisation injected by spring as request scope variable...
Not now,  but would be worth looking into that in future. Maybe I will look into it.

Reply:
happy to look into this in the future also. Found a good little guide on spring bean scopes

Reply:
Opened PR with that :) https://github.com/hmcts/fpl-ccd-configuration/pull/749